cmd/xl-sets.go

```diff
@@ -0,0 +1,1101 @@
+	if len(s.sets) == 1 {
+		return s.sets[0].DeleteBucket(bucket)
+	}
+
+	g := errgroup.WithNErrs(len(s.sets))
+
+	// Delete buckets in parallel across all sets.
+	for index := range s.sets {
+		index := index
+		g.Go(func() error {
+			return s.sets[index].DeleteBucket(bucket)
+		}, index)
+	}
+
+	errs := g.Wait()
+	// For any failure, we undo all the delete buckets operation
+	// by creating all the buckets.
+	for _, err := range errs {
+		if err != nil {
+			undoDeleteBucketSets(bucket, s.sets, errs)
+			return err
+		}
+	}
+
+	// Success.
+	return nil
+}
+
+// This function is used to undo a successful DeleteBucket operation.
+func undoDeleteBucketSets(bucket string, sets []*xlObjects, errs []error) {
+	g := errgroup.WithNErrs(len(sets))
+
+	// Undo previous delete bucket on all underlying sets.
+	for index := range sets {
+		index := index
+		if errs[index] == nil {
+			g.Go(func() error {
+				return sets[index].MakeBucketWithLocation(bucket, "")
+			}, index)
+		}
+	}
+
+	g.Wait()
+}
+
+// List all buckets from one of the set, we are not doing merge
+// sort here just for simplification. As per design it is assumed
+// that all buckets are present on all sets.
+func (s *xlSets) ListBuckets() (buckets []BucketInfo, err error) {
+	// Always lists from the same set signified by the empty string.
+	return s.getHashedSet("").ListBuckets()
+}
+
+// --- Object Operations ---
+
+// GetObject - reads an object from the hashedSet based on the object name.
+func (s *xlSets) GetObject(bucket, object string, startOffset int64, length int64, writer io.Writer, etag string) error {
+	return s.getHashedSet(object).GetObject(bucket, object, startOffset, length, writer, etag)
+}
+
+// PutObject - writes an object to hashedSet based on the object name.
+func (s *xlSets) PutObject(bucket string, object string, data *hash.Reader, metadata map[string]string) (objInfo ObjectInfo, err error) {
+	return s.getHashedSet(object).PutObject(bucket, object, data, metadata)
+}
+
+// GetObjectInfo - reads object metadata from the hashedSet based on the object name.
+func (s *xlSets) GetObjectInfo(bucket, object string) (objInfo ObjectInfo, err error) {
+	return s.getHashedSet(object).GetObjectInfo(bucket, object)
+}
+
+// DeleteObject - deletes an object from the hashedSet based on the object name.
+func (s *xlSets) DeleteObject(bucket string, object string) (err error) {
+	return s.getHashedSet(object).DeleteObject(bucket, object)
+}
+
+// CopyObject - copies objects from one hashedSet to another hashedSet, on server side.
+func (s *xlSets) CopyObject(srcBucket, srcObject, destBucket, destObject string, metadata map[string]string, srcEtag string) (objInfo ObjectInfo, err error) {
+	if len(s.sets) == 1 {
+		return s.sets[0].CopyObject(srcBucket, srcObject, destBucket, destObject, metadata, srcEtag)
+	}
+
+	srcSet := s.getHashedSet(srcObject)
+	destSet := s.getHashedSet(destObject)
+
+	objInfo, err = srcSet.GetObjectInfo(srcBucket, srcObject)
```

Comment:
better make sure that srcEtag has not changed here.

Reply:
why such a premature step? srcETag is passed into all relevant I/O calls anyways.

Reply:
That's true